src/Ocelot/Requester/MessageInvokerPool.cs

```diff
@@ -0,0 +1,108 @@
+        // and the QoS Options properties can't be changed after the route is created,
+        // we don't need to use the timeout value as part of the cache key.
+        return _handlersPool.GetOrAdd(
+            new MessageInvokerCacheKey(downstreamRoute),
+            cacheKey => new Lazy<HttpMessageInvoker>(() => CreateMessageInvoker(cacheKey.DownstreamRoute))
+        ).Value;
+    }
+
+    public void Clear() => _handlersPool.Clear();
+
+    private HttpMessageInvoker CreateMessageInvoker(DownstreamRoute downstreamRoute)
+    {
+        var baseHandler = CreateHandler(downstreamRoute);
+        var handlers = _handlerFactory.Get(downstreamRoute).Data;
+        handlers.Reverse();
+
+        foreach (var delegatingHandler in handlers.Select(handler => handler()))
+        {
+            delegatingHandler.InnerHandler = baseHandler;
+            baseHandler = delegatingHandler;
+        }
+
+        // Adding timeout handler to the top of the chain.
+        // It's standard behavior to throw TimeoutException after the defined timeout (90 seconds by default)
+        var timeoutHandler = new TimeoutDelegatingHandler(downstreamRoute.QosOptions.TimeoutValue == 0
+            ? TimeSpan.FromSeconds(DefaultRequestTimeoutSeconds)
+            : TimeSpan.FromMilliseconds(downstreamRoute.QosOptions.TimeoutValue))
+        {
+            InnerHandler = baseHandler,
+        };
+
+        return new HttpMessageInvoker(timeoutHandler, true);
+    }
+
+    private HttpMessageHandler CreateHandler(DownstreamRoute downstreamRoute)
+    {
+        var handler = new SocketsHttpHandler
+        {
+            AllowAutoRedirect = downstreamRoute.HttpHandlerOptions.AllowAutoRedirect,
+            UseCookies = downstreamRoute.HttpHandlerOptions.UseCookieContainer,
+            UseProxy = downstreamRoute.HttpHandlerOptions.UseProxy,
+            MaxConnectionsPerServer = downstreamRoute.HttpHandlerOptions.MaxConnectionsPerServer,
+            PooledConnectionLifetime = downstreamRoute.HttpHandlerOptions.PooledConnectionLifeTime,
+        };
+
+        if (downstreamRoute.HttpHandlerOptions.UseCookieContainer)
+        {
+            handler.CookieContainer = new CookieContainer();
+        }
+
+        if (!downstreamRoute.DangerousAcceptAnyServerCertificateValidator)
+        {
+            return handler;
+        }
+
+        handler.SslOptions = new SslClientAuthenticationOptions
+        {
+            RemoteCertificateValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true,
+        };
+
+        _logger.LogWarning(() =>
+            $"You have ignored all SSL warnings by using DangerousAcceptAnyServerCertificateValidator for this DownstreamRoute, UpstreamPathTemplate: {downstreamRoute.UpstreamPathTemplate}, DownstreamPathTemplate: {downstreamRoute.DownstreamPathTemplate}");
+
+        return handler;
+    }
+
+    private readonly struct MessageInvokerCacheKey(DownstreamRoute downstreamRoute) : IEquatable<MessageInvokerCacheKey>
```

Comment:
Why `private`? Because pooler will work behind the scenes?
Maybe `public` in separate .cs file?!...

Reply:
Yes, we could, here it's a struct. I followed an example given by @RaynaldM But, yeah, we could change, easier to unit test.

Reply:
If it is not hard refactoring, you could move this to separate file while working on tests...